From 7ec02d0bd31f2b4e870f060ba2e1d6c4d0c8ff72 Mon Sep 17 00:00:00 2001
From: Brett Mastbergen <bmastbergen@untangle.com>
Date: Mon, 13 Jan 2020 16:33:01 -0500
Subject: [PATCH 4/8] untangle: iptables-tune

This adds the "tune" table in netfilter which comes at the very end
which is used for various tasks in our networking setup.
---
 include/net/netns/ipv4.h            |   1 +
 include/uapi/linux/netfilter_ipv4.h |   3 +-
 net/ipv4/netfilter/Kconfig          |  12 ++++
 net/ipv4/netfilter/Makefile         |   1 +
 net/ipv4/netfilter/iptable_tune.c   | 102 ++++++++++++++++++++++++++++
 5 files changed, 118 insertions(+), 1 deletion(-)
 create mode 100644 net/ipv4/netfilter/iptable_tune.c

diff --git a/include/net/netns/ipv4.h b/include/net/netns/ipv4.h
index 8e4fcac..921ebc2 100644
--- a/include/net/netns/ipv4.h
+++ b/include/net/netns/ipv4.h
@@ -77,6 +77,7 @@ struct netns_ipv4 {
 	struct xt_table		*iptable_filter;
 	struct xt_table		*iptable_mangle;
 	struct xt_table		*iptable_raw;
+	struct xt_table		*iptable_tune;
 	struct xt_table		*arptable_filter;
 #ifdef CONFIG_SECURITY
 	struct xt_table		*iptable_security;
diff --git a/include/uapi/linux/netfilter_ipv4.h b/include/uapi/linux/netfilter_ipv4.h
index 155e77d..8366654 100644
--- a/include/uapi/linux/netfilter_ipv4.h
+++ b/include/uapi/linux/netfilter_ipv4.h
@@ -41,7 +41,8 @@ enum nf_ip_hook_priorities {
 	NF_IP_PRI_NAT_SRC = 100,
 	NF_IP_PRI_SELINUX_LAST = 225,
 	NF_IP_PRI_CONNTRACK_HELPER = 300,
-	NF_IP_PRI_CONNTRACK_CONFIRM = INT_MAX,
+	NF_IP_PRI_CONNTRACK_CONFIRM = INT_MAX - 10,
+	NF_IP_PRI_TUNE = INT_MAX - 5,
 	NF_IP_PRI_LAST = INT_MAX,
 };
 
diff --git a/net/ipv4/netfilter/Kconfig b/net/ipv4/netfilter/Kconfig
index a2f4f89..f13dc78 100644
--- a/net/ipv4/netfilter/Kconfig
+++ b/net/ipv4/netfilter/Kconfig
@@ -358,3 +358,15 @@ endif # IP_NF_ARPTABLES
 
 endmenu
 
+# tune + specific targets
+config IP_NF_TUNE
+	tristate  'tune table support (adds a chain after POSTROUTING nat)'
+	depends on IP_NF_IPTABLES
+	help
+	  This option adds a `tune' table to iptables. This table is the very
+	  last in the netfilter framework and hooks in at the POSTROUTING
+	  chain.
+	
+	  If you want to compile it as a module, say M here and read
+	  <file:Documentation/modules.txt>.  If unsure, say `N'.
+
diff --git a/net/ipv4/netfilter/Makefile b/net/ipv4/netfilter/Makefile
index 7c497c7..25924fe 100644
--- a/net/ipv4/netfilter/Makefile
+++ b/net/ipv4/netfilter/Makefile
@@ -39,6 +39,7 @@ obj-$(CONFIG_IP_NF_FILTER) += iptable_filter.o
 obj-$(CONFIG_IP_NF_MANGLE) += iptable_mangle.o
 obj-$(CONFIG_IP_NF_NAT) += iptable_nat.o
 obj-$(CONFIG_IP_NF_RAW) += iptable_raw.o
+obj-$(CONFIG_IP_NF_TUNE) += iptable_tune.o
 obj-$(CONFIG_IP_NF_SECURITY) += iptable_security.o
 
 # matches
diff --git a/net/ipv4/netfilter/iptable_tune.c b/net/ipv4/netfilter/iptable_tune.c
new file mode 100644
index 0000000..c976e34
--- /dev/null
+++ b/net/ipv4/netfilter/iptable_tune.c
@@ -0,0 +1,102 @@
+/* 
+ * 
+ * (C) 2015 David Han <dhan@untangle.com>
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/netfilter.h>
+#include <linux/netfilter_ipv4.h>
+#include <linux/netfilter_ipv4/ip_tables.h>
+#include <linux/ip.h>
+#include <net/ip.h>
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Untangle Inc. www.untangle.com");
+MODULE_DESCRIPTION("iptables tune table");
+
+#define TUNE_VALID_HOOKS ((1 << NF_INET_POST_ROUTING) | (1 << NF_INET_LOCAL_OUT))
+
+static int __net_init iptable_tune_table_init(struct net *net);
+
+static const struct xt_table packet_tuner = {
+	.name		= "tune",
+	.valid_hooks	= TUNE_VALID_HOOKS,
+	.me		= THIS_MODULE,
+	.af		= NFPROTO_IPV4,
+	.priority = NF_IP_PRI_TUNE,
+	.table_init = iptable_tune_table_init,
+};
+
+static unsigned int
+iptable_tune_hook(void *priv, struct sk_buff *skb,
+		    const struct nf_hook_state *state)
+{
+	return ipt_do_table(skb, state, state->net->ipv4.iptable_tune);
+}
+
+static struct nf_hook_ops *tune_ops __read_mostly;
+
+static int __net_init iptable_tune_table_init(struct net *net)
+{
+	struct ipt_replace *repl;
+	int err;
+
+	if (net->ipv4.iptable_tune)
+		return 0;
+
+	repl = ipt_alloc_initial_table(&packet_tuner);
+	if (repl == NULL)
+		return -ENOMEM;
+
+	err = ipt_register_table(net, &packet_tuner, repl,
+				 tune_ops,
+				 &net->ipv4.iptable_tune);
+	kfree(repl);
+	return err;
+}
+
+static void __net_exit iptable_tune_net_exit(struct net *net)
+{
+	if (!net->ipv4.iptable_tune)
+		return;
+	ipt_unregister_table(net, net->ipv4.iptable_tune, tune_ops);
+	net->ipv4.iptable_tune = NULL;
+}
+
+static struct pernet_operations iptable_tune_net_ops = {
+	.exit	= iptable_tune_net_exit,
+};
+
+static int __init iptable_tune_init(void)
+{
+	int ret;
+
+	tune_ops = xt_hook_ops_alloc(&packet_tuner, iptable_tune_hook);
+	if (IS_ERR(tune_ops)) {
+		return PTR_ERR(tune_ops);
+	}
+
+	ret = register_pernet_subsys(&iptable_tune_net_ops);
+	if (ret < 0) {
+ 		kfree(tune_ops);
+		return ret;
+	}
+
+	ret = iptable_tune_table_init(&init_net);
+	if (ret) {
+		unregister_pernet_subsys(&iptable_tune_net_ops);
+		kfree(tune_ops);
+	}
+
+	return ret;
+}
+
+static void __exit iptable_tune_exit(void)
+{
+	unregister_pernet_subsys(&iptable_tune_net_ops);
+	kfree(tune_ops);
+}
+
+module_init(iptable_tune_init);
+module_exit(iptable_tune_exit);
-- 
2.32.0

